doc/qbk/23_sql_formatting_advanced.qbk

```diff
@@ -0,0 +1,181 @@
+instead.
+
+[reflink basic_format_context] contains an error code that is set when formatting
+a value fails. This is called the ['error state], and can be queried using [refmem basic_format_context error_state].
+When [refmem basic_format_context get] is called (after all individual values have been formatted),
+the error state is checked. The `system::result` returned by `get` will contain the error
+state if it was set, or the generated query if it was not:
+
+[sql_formatting_no_exceptions]
+
+Rationale: the error state mechanism makes composing formatters easier,
+as the error state is checked only once.
+
+Errors caused by invalid format strings are also reported using this mechanism.
+
+
+
+
+
+[heading:format_options Format options and character set tracking]
+
+MySQL has many configuration options that affect its syntax. There are two options 
+that formatting functions need to know in order to work:
+
+* Whether the backslash character represents an escape sequence or not. By default it does,
+  but this can be disabled dynamically by setting the
+  [@https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_no_backslash_escapes NO_BACKSLASH_ESCAPES] SQL mode.
+  This is tracked by [reflink any_connection] automatically (see [refmem any_connection backslash_escapes]).
+* The connection's [*current character set]. This determines which multi-byte sequences are valid,
+  and is required to iterate and escape the string. The current character set is tracked
+  by connections as far as possible, but deficiencies in the protocol create cases where the character
+  set may not be known to the client. The current character set can be accessed using
+  [refmem any_connection current_character_set].
+
+[refmem any_connection format_opts] is a convenience function that returns a
+[link mysql.error_handling.system_result `boost::system::result`]`<`[reflink format_options]`>`.
+If the connection could not determine the current character set, the result will contain an error.
+For a reference on how character set tracking works, please read [link mysql.charsets.tracking this section].
+  
+[warning
+    Passing an incorrect `format_options` value to formatting functions may cause
+    escaping to generate incorrect values, which may generate vulnerabilities.
+    Stay safe and always use [refmem any_connection format_opts] instead of
+    hand-crafting `format_options` values.
+]
+
+
+
+[heading Custom string types]
+
+[reflink format_sql_to] can be used with string types that are not `std::string`,
+as long as they satisfy the [reflink OutputString] concept. This includes
+strings with custom allocators (like `std::pmr::string`) and `boost::static_string`.
+You need to use [reflink basic_format_context], specifying the string type:
+
+[sql_formatting_custom_string]
+
+
+
+
+
+[heading Re-using string memory]
+
+You can pass a string value to the context's constructor, to re-use memory:
```

Comment:
maybe mention how the value is being used so that the user can get a mental model of how to structure a reuse scenario.

for example, is the input captured by reference, and used. or maybe is it moved into the constructor, and then moved out when calling `.get() `.